staging/src/slime.io/slime/modules/meshregistry/pkg/source/zookeeper/watching.go

```diff
@@ -32,228 +34,445 @@ func (s *Source) ServiceNodeDelete(path string) {
-	sort.Slice(endpoints, func(i, j int) bool {
-		if endpoints[i].Address == endpoints[j].Address {
-			return labelToSstring(endpoints[i].Labels) < labelToSstring(endpoints[j].Labels)
+type ServiceWatcher struct {
+	ctx context.Context
+
+	conn               *atomic.Value
+	rootPath           string
+	endpointUpdateFunc func([]string, []string, string)
+	serviceDeleteFunc  func(string)
+	gatewatModel       bool
+
+	svcs []string
+
+	workers []*worker
+
+	initLock               sync.Mutex
+	initWait               sync.WaitGroup
+	initCnt, initThreshold int
+}
+
+// block until initialized
+func (sw *ServiceWatcher) Start(ctx context.Context) {
+	sw.ctx = ctx
+	initCh := make(chan struct{})
+	go func() {
+		firstLoop := true
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			svcs, _, e, err := sw.conn.Load().(*zk.Conn).ChildrenW(sw.rootPath)
+			if err != nil {
+				// retry and mark ready
+				firstLoop = false
+				close(initCh)
+				continue
+			}
+			if firstLoop {
+				firstLoop = false
+				// For the zk watching mode, we initialize according to the number of services
+				// obtained after registering the watch root node for the first time.
+				// When the ep watchers for these services start up and perform their first processing,
+				// the zk source is marked as ready.
+				sw.initThreshold = len(svcs)
+				sw.initWait.Add(sw.initThreshold)
+				go func() {
+					sw.initWait.Wait()
+					close(initCh)
+				}()
+			}
+			sw.handleSvcs(svcs)
+			<-e
 		}
-		return endpoints[i].Address < endpoints[j].Address
-	})
+	}()
+	<-initCh
 }
 
-func verifyPath(path string, isInstance bool) bool {
-	ss := strings.Split(path, "/")
-	if len(ss) < 2 {
-		scope.Errorf("Invalid watch path")
-		return false
+func (sw *ServiceWatcher) handleSvcs(svcs []string) {
+	var oldSvcs []string
+	sw.svcs, oldSvcs = svcs, sw.svcs
+	scope.Debugf("service watcher handle event with current services: %v, last services: %v", sw.svcs, oldSvcs)
+	deleted, created := calculateDiff(oldSvcs, sw.svcs)
+	scope.Debugf("service watcher calculate diff with delete: %v, create: %v", deleted, created)
+	// dispatch deleted service first
+	for _, d := range deleted {
+		sw.dispatch(ServiceEvent{
+			path:  filepath.Join(sw.rootPath, d),
+			etype: EventTypeDelete,
+		})
+	}
+	for _, c := range created {
+		sw.dispatch(ServiceEvent{
+			path:  filepath.Join(sw.rootPath, c),
+			etype: EventTypeCreate,
+		})
 	}
-	if isInstance {
-		if ss[len(ss)-1] != ConsumerNode && ss[len(ss)-1] != ProviderNode {
-			scope.Errorf("Invalid watch path")
-			return false
+}
+
+func (sw *ServiceWatcher) dispatch(e ServiceEvent) {
+	workerIdx := fnv32(e.path) % uint32(len(sw.workers))
+	if sw.workers[workerIdx] == nil {
+		w := NewServiceWorker(sw.conn, sw.endpointUpdateFunc, sw.serviceDeleteFunc, sw.gatewatModel, sw.initCallbackFactory)
+		w.Start(sw.ctx)
+		sw.workers[workerIdx] = w
+	}
+	sw.workers[workerIdx].HandleEvent(e)
```

Comment:
Not very well understood

Reply:
We defer worker initialization to avoid creating unused workers